staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+		w.setBookmarkAfterResourceVersion(10)
+		return w
+	}
+
+	clock := testingclock.NewFakeClock(time.Now())
+	target := newTimeBucketWatchers(clock, defaultBookmarkFrequency)
+	if !target.addWatcher(newWatcher("1", clock.Now().Add(2*time.Minute))) {
+		t.Fatal("failed adding an even to the watcher")
+	}
+
+	// the watcher is immediately expired
+	ret := target.popExpiredWatchers()
+	if len(ret) != 1 || len(ret[0]) != 1 {
+		t.Fatalf("expected only one watcher to be expired")
+	}
+	if !target.addWatcher(ret[0][0]) {
+		t.Fatal("failed adding an even to the watcher")
+	}
+
+	// after one second time the watcher is still expired
+	clock.Step(1 * time.Second)
+	ret = target.popExpiredWatchers()
+	if len(ret) != 1 || len(ret[0]) != 1 {
+		t.Fatalf("expected only one watcher to be expired")
+	}
+	if !target.addWatcher(ret[0][0]) {
+		t.Fatal("failed adding an even to the watcher")
+	}
+
+	// after 29 seconds the watcher is still expired
+	clock.Step(29 * time.Second)
+	ret = target.popExpiredWatchers()
+	if len(ret) != 1 || len(ret[0]) != 1 {
+		t.Fatalf("expected only one watcher to be expired")
+	}
+
+	// after confirming the watcher is not expired immediately
+	ret[0][0].markBookmarkAfterResourceVersionAsReceived(&watchCacheEvent{Type: watch.Bookmark, ResourceVersion: 10, Object: &v1.Pod{}})
+	if !target.addWatcher(ret[0][0]) {
+		t.Fatal("failed adding an even to the watcher")
+	}
+	clock.Step(30 * time.Second)
+	ret = target.popExpiredWatchers()
+	if len(ret) != 0 {
+		t.Fatalf("didn't expect any watchers to be expired")
+	}
+
+	clock.Step(30 * time.Second)
+	ret = target.popExpiredWatchers()
+	if len(ret) != 1 || len(ret[0]) != 1 {
+		t.Fatalf("expected only one watcher to be expired")
+	}
+}
+
+func TestCacheWatchSemantics(t *testing.T) {
+	makePod := func(rv uint64) *example.Pod {
+		return &example.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            fmt.Sprintf("pod-%d", rv),
+				Namespace:       "ns",
+				ResourceVersion: fmt.Sprintf("%d", rv),
+				Annotations:     map[string]string{},
+			},
+		}
+	}
+
+	scenarios := []struct {
+		name string
+		run  func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string))
+	}{
+		{
+			name: "allowWatchBookmarks=true, sendInitialEvents=true, RV=unset",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// before starting a new watch set a storage RV to some future value
+				setStorageListMetaRV("102")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{Predicate: func() storage.SelectionPredicate { p := storage.Everything; p.AllowWatchBookmarks = true; return p }(), SendInitialEvents: func() *bool { b := true; return &b }()})
```

Comment:
Can you avoid those long line? Also let's do instead:
```
opts := storage.ListOptions{
  Labels: labels.Everything(),
  Fields: fields.Everything(),
  AllowWatchBookmarks: true.
  SendInitialEvents: &trueValue,
}
w, err := cacher.Watch(...)
```

[and define trueValue and falseValue above]